cranelift/codegen/src/ir/pcc.rs

```diff
@@ -0,0 +1,347 @@
+                },
+            ) if *bw_max >= pointer_width && add_width >= *bw_max => {
+                let computed_region = MemoryRegion {
+                    max: region.max.checked_sub(*max)?,
+                };
+                Some(Fact::PointsTo {
+                    region: computed_region,
+                })
+            }
+
+            _ => None,
+        }
+    }
+
+    /// Computes the `uextend` of a value with the given facts.
+    pub fn uextend(&self, from_width: u16, to_width: u16) -> Option<Self> {
+        match self {
+            // If we have a defined value in bits 0..bit_width, and we
+            // are filling zeroes into from_bits..to_bits, and
+            // bit_width and from_bits are exactly contiguous, then we
+            // have defined values in 0..to_bits (and because this is
+            // a zero-extend, the max value is the same).
+            Fact::ValueMax { bit_width, max } if *bit_width == from_width => Some(Fact::ValueMax {
+                bit_width: to_width,
+                max: *max,
+            }),
+            _ => None,
+        }
+    }
+
+    /// Computes the `sextend` of a value with the given facts.
+    pub fn sextend(&self, from_width: u16, to_width: u16) -> Option<Self> {
+        match self {
+            // If we have a defined value in bits 0..bit_width, and
+            // the MSB w.r.t. `from_width` is *not* set, then we can
+            // do the same as `uextend`.
+            Fact::ValueMax { bit_width, max }
+                if *bit_width == from_width && (*max & (1 << (*bit_width - 1)) == 0) =>
+            {
+                self.uextend(from_width, to_width)
+            }
+            _ => None,
+        }
+    }
+
+    /// Scales a value with a fact by a known constant.
+    pub fn scale(&self, width: u16, factor: u32) -> Option<Self> {
+        match self {
+            Fact::ValueMax { bit_width, max } if *bit_width == width => {
+                let max = max.checked_mul(factor as u64)?;
+                if *bit_width < 64 && max > ((1 << width) - 1) {
+                    return None;
+                }
+                Some(Fact::ValueMax {
+                    bit_width: *bit_width,
+                    max,
+                })
+            }
+            _ => None,
+        }
+    }
+
+    /// Offsets a value with a fact by a known amount.
+    pub fn offset(&self, width: u16, offset: i64) -> Option<Self> {
+        // If we eventually support two-sided ranges, we can
+        // represent (0..n) + m -> ((0+m)..(n+m)). However,
+        // right now, all ranges start with zero, so any
+        // negative offset could underflow, and removes all
+        // claims of constrained range.
+        if offset < 0 {
+            return None;
+        }
+        let offset = u64::try_from(offset).unwrap();
+
+        match self {
+            Fact::ValueMax { bit_width, max } if *bit_width == width => {
+                let max = max.checked_add(offset).unwrap();
```

Comment:
Wait why is it okay to unwrap here? Shouldn't this use `?` to propagate?

Reply:
Good point; now this returns the "minimal fact" (the range of the whole bitwidth) instead if this overflows.